File: app/src/main/java/im/fdx/v2ex/network/HttpHelper.java
package im.fdx.v2ex.network;

import com.franmontiel.persistentcookiejar.PersistentCookieJar;
import com.franmontiel.persistentcookiejar.cache.SetCookieCache;
import com.franmontiel.persistentcookiejar.persistence.SharedPrefsCookiePersistor;
import com.readystatesoftware.chuck.ChuckInterceptor;

import im.fdx.v2ex.MyApp;
import im.fdx.v2ex.network.cookie.CookiePersistor;
import im.fdx.v2ex.network.cookie.MyCookieJar;
import im.fdx.v2ex.network.cookie.SharedPrefsPersistor;
import okhttp3.Headers;
import okhttp3.OkHttpClient;

/**
 * Created by fdx on 2016/11/20.
 * fdx will maintain it
 */

public class HttpHelper {

    public static final int REQUEST_SIGNUP = 0;
    private static PersistentCookieJar cookieJar = new PersistentCookieJar(new SetCookieCache(), new SharedPrefsCookiePersistor(MyApp.getInstance().getApplicationContext()));
    public static final MyCookieJar myCookieJar = new MyCookieJar(new SharedPrefsPersistor(MyApp.getInstance().getApplicationContext()));
    public static final OkHttpClient OK_CLIENT = new OkHttpClient().newBuilder()
//            .addInterceptor(new HttpLoggingInterceptor())
//            .connectTimeout(10, TimeUnit.SECONDS)
//            .writeTimeout(10, TimeUnit.SECONDS)
//            .readTimeout(30, TimeUnit.SECONDS)
            .followRedirects(false)  //禁止重定向
            .addInterceptor(new ChuckInterceptor(MyApp.getInstance().getApplicationContext()))//好东西，查看Okhttp数据
            .cookieJar(myCookieJar)
//            .cookieJar(cookieJar)
            .build();
    public static Headers baseHeaders = new Headers.Builder()
            .add("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8")
            .add("Accept-Charset", "utf-8, iso-8859-1, utf-16, *;q=0.7")
            .add("Accept-Language", "zh-CN,zh;q=0.8,en;q=0.6")
            .add("Host", "www.v2ex.com")
            .add("Cache-Control", "max-age=0")
//            .add("X-Requested-With", "com.android.browser")
//            .add("User-Agent", "Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3013.3 Mobile Safari/537.36");
            .add("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36" +
                    " (KHTML, like Gecko) Chrome/58.0.3013.3 Safari/537.36")
            .build();


}

File: app/src/main/java/im/fdx/v2ex/network/HttpHelper.kt
package im.fdx.v2ex.network

import android.util.Log
import com.chuckerteam.chucker.api.ChuckerInterceptor
import com.google.firebase.crashlytics.FirebaseCrashlytics
import im.fdx.v2ex.MyApp
import im.fdx.v2ex.network.cookie.MyCookieJar
import im.fdx.v2ex.network.cookie.SharedPrefsPersistor
import im.fdx.v2ex.utils.extensions.logd
import im.fdx.v2ex.utils.extensions.logi
import okhttp3.*
import okhttp3.logging.HttpLoggingInterceptor
import java.io.IOException
import java.util.concurrent.TimeUnit
import java.util.logging.Level
import java.util.logging.Logger

/**
 * Created by fdx on 2016/11/20.
 * fdx will maintain it
 */

object HttpHelper {

    val cookiePersistor = SharedPrefsPersistor(MyApp.get())
    val myCookieJar: MyCookieJar = MyCookieJar(cookiePersistor)

    val OK_CLIENT: OkHttpClient = OkHttpClient().newBuilder()
//                        .addInterceptor(HttpLoggingInterceptor())
                        .connectTimeout(10, TimeUnit.SECONDS)
                        .writeTimeout(10, TimeUnit.SECONDS)
                        .readTimeout(30, TimeUnit.SECONDS)
            .followRedirects(false)  //禁止重定向
            .addNetworkInterceptor(
                        HttpLoggingInterceptor(object : HttpLoggingInterceptor.Logger {
                            override fun log(message: String) {
                                logd("okhttp: $message")
                            }
                        }).apply { level = HttpLoggingInterceptor.Level.HEADERS }
                    )
            .addInterceptor(ChuckerInterceptor(MyApp.get()))//好东西，查看Okhttp数据
            .addInterceptor { chain ->
                val request = chain.request()
                        .newBuilder()
                        .header("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9")
                        .header("Accept-Charset", "utf-8, iso-8859-1, utf-16, *;q=0.7")
//                        .header("Accept-Encoding", "gzip, deflate, br")
                        .header("Accept-Language", "zh-CN,zh;q=0.9,en;q=0.8,zh-TW;q=0.7")
                        .header("Host", "www.v2ex.com")
                        .header("Authority", "v2ex.com")
                        .header("Cache-Control", "max-age=0")
//                        .header("User-Agent", "Mozilla/5.0 (Linux; Android 8.0; Pixel 2 Build/OPD3.170816.012) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.116 Mobile Safari/537.36")
                        //换回PC端的User-Agent, 因为一些奇怪的问题
                        .header("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.82 Safari/537.36")
                        .build()
                chain.proceed(request)
            }
            .cookieJar(myCookieJar)
            .build()

}

fun vCall(url: String): Call {
    Log.d("fdx", "vCall: $url ")
    return HttpHelper.OK_CLIENT.newCall(Request.Builder().url(url).build())
}


fun Call.start(callback: Callback) {
    enqueue(object : Callback {
        override fun onFailure(call: Call, e: IOException) {
            try {
                callback.onFailure(call, e)
            } catch (e2: Exception) {
                e2.printStackTrace()
            }
        }

        override fun onResponse(call: Call, response: Response) {
            try {
                callback.onResponse(call, response)
            } catch (e: Exception) {
                FirebaseCrashlytics.getInstance().recordException(e)
                e.printStackTrace()
            }
        }
    })
}


@Deprecated("暂时还用不上，简化代码之用")
fun Call.start(onResp: (Call, Response) -> Unit, onFail: (Call, IOException)-> Unit){
    val callback = object: Callback {
        override fun onResponse(call: Call, response: Response) {
            onResp(call, response)
        }

        override fun onFailure(call: Call, e: IOException) {
            onFail(call , e)
        }
    }
    enqueue(object : Callback {
        override fun onFailure(call: Call, e: IOException) {
            try {
                callback.onFailure(call, e)
            } catch (e2: Exception) {
                FirebaseCrashlytics.getInstance().recordException(e2)
                e2.printStackTrace()
            }
        }

        override fun onResponse(call: Call, response: Response) {
            try {
                callback.onResponse(call, response)
            } catch (e: Exception) {
                FirebaseCrashlytics.getInstance().recordException(e)
                e.printStackTrace()
            }
        }
    })
}



